esdt-safe/src/lib.rs

```diff
@@ -0,0 +1,309 @@
+            }
+            None => Ok(OptionalResult::None),
+        }
+    }
+
+    #[endpoint(setTransactionStatus)]
+    fn set_transaction_status_endpoint(
+        &self,
+        sender: Address,
+        nonce: usize,
+        transaction_status: TransactionStatus,
+    ) -> SCResult<()> {
+        only_owner!(self, "only owner may call this function");
+
+        require!(
+            self.get_transaction_status(&sender, nonce) == TransactionStatus::InProgress,
+            "Transaction has to be executed first"
+        );
+
+        match transaction_status {
+            TransactionStatus::Executed => {
+                self.set_transaction_status(&sender, nonce, TransactionStatus::Executed);
+
+                let tx = self.transactions_by_nonce(&sender).get(nonce);
+                self.burn_esdt_token(&tx.token_identifier, &tx.amount);
+            }
+            TransactionStatus::Rejected => {
+                self.set_transaction_status(&sender, nonce, TransactionStatus::Rejected);
+
+                let tx = self.transactions_by_nonce(&sender).get(nonce);
+                self.refund_esdt_token(&tx.from, &tx.token_identifier, &tx.amount);
+            }
+            _ => return sc_error!("Transaction status may only be set to Executed or Rejected"),
+        }
+
+        self.set_transaction_status(&sender, nonce, transaction_status);
+
+        Ok(())
+    }
+
+    #[endpoint]
+    fn claim(&self) -> SCResult<()> {
+        only_owner!(self, "only owner may call this function");
+
+        self.send().direct_egld(
+            &self.get_caller(),
+            &self.get_claimable_transaction_fee(),
+            b"claim",
+        );
+
+        self.set_claimable_transaction_fee(&BigUint::zero());
+
+        Ok(())
+    }
+
+    // endpoints
+
+    #[payable("EGLD")]
+    #[endpoint(depositEgldForTransactionFee)]
+    fn deposit_egld_for_transaction_fee(&self, #[payment] payment: BigUint) {
+        let caller = self.get_caller();
+        let mut caller_deposit = self.get_deposit(&caller);
+        caller_deposit += payment;
+        self.set_deposit(&caller, &caller_deposit);
+    }
+
+    /// amount argument is optional, defaults to max possible if not provided
+    #[endpoint(whithdrawDeposit)]
+    fn whithdraw_deposit(&self, #[var_args] opt_amount: OptionalArg<BigUint>) -> SCResult<()> {
+        let caller = self.get_caller();
+        let caller_deposit = self.get_deposit(&caller);
+        let amount = match opt_amount {
+            OptionalArg::Some(value) => value,
+            OptionalArg::None => caller_deposit.clone(),
+        };
+
+        require!(amount <= caller_deposit, "Trying to whithdraw too much");
+
+        let deposit_remaining = &caller_deposit - &amount;
+        self.send().direct_egld(&caller, &amount, b"whitdrawal");
+        self.set_deposit(&caller, &deposit_remaining);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint(createTransaction)]
+    fn create_transaction(
```

Comment:
what does this do ? the Safe contract only needs to lock, burn and send back tokens if the bridge failed. BURN and SEND BACK can be called by whitelisted address | smart contract owner only.

Reply:
This endpoint is called by accounts that want to transfer to another chain. A transaction struct is created and saved with the required data (sender, destination, ESDT token used and amount paid), which is saved as a "pending" transaction. The tokens are locked in the contract after this.